pilot/adapter/config/file/monitor_test.go

```diff
@@ -0,0 +1,259 @@
+	controller     model.ConfigStoreCache
+	controllerStop chan struct{}
+}
+
+func (cm *controllerManager) setup(eventCh chan event) {
+	cm.controllerStop = make(chan struct{})
+	store := memory.Make(model.IstioConfigTypes)
+	cm.controller = memory.NewControllerWithBufferSize(store, 100)
+
+	// Register changes to the repository
+	for _, s := range model.IstioConfigTypes.Types() {
+		cm.controller.RegisterEventHandler(s, func(config model.Config, ev model.Event) {
+			eventCh <- event{
+				config: config,
+				event:  ev,
+			}
+		})
+	}
+
+	// Run the controller.
+	go cm.controller.Run(cm.controllerStop)
+}
+
+func (cm *controllerManager) teardown() {
+	close(cm.controllerStop)
+}
+
+type env struct {
+	controllerManager controllerManager
+	monitorCtx        context.Context
+	monitorCancel     context.CancelFunc
+	eventCh           chan event
+	fsroot            string
+	monitor           *file.Monitor
+}
+
+func (e *env) setup(tempDir string) {
+	e.eventCh = make(chan event)
+
+	// Create and setup the controller.
+	e.controllerManager = controllerManager{}
+	e.controllerManager.setup(e.eventCh)
+
+	e.monitorCtx, e.monitorCancel = context.WithCancel(context.Background())
+	e.fsroot = tempDir
+
+	// Make all of the updates go through the controller so that the events are triggered.
+	e.monitor = file.NewMonitor(e.controllerManager.controller, e.fsroot, nil)
+	e.monitor.Start(e.monitorCtx)
+}
+
+func (e *env) teardown() {
+	e.monitorCancel()
+	e.controllerManager.teardown()
+
+	// Remove the temp dir.
+	os.RemoveAll(e.fsroot)
+}
+
+func (e *env) watch() event {
+	return <-e.eventCh
+}
+
+func (e *env) watchFor(eventType model.Event) event {
+	for {
+		if ev := e.watch(); ev.event == eventType {
+			return ev
+		}
+	}
+}
+
+func copyTestFile(fileName string, toDir string) error {
+	return copyFile("testdata/"+fileName, toDir+"/"+fileName)
+}
+
+func copyFile(src, dst string) error {
+	data, err := ioutil.ReadFile(src)
+	if err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(dst, data, os.ModePerm)
+}
+
+func parseConfig(path string) (*model.Config, error) {
+	configs, err := file.ParseYamlFile(path)
+	if err != nil {
+		return nil, err
+	}
+	return configs[0], nil
+}
+
+func TestAdd(t *testing.T) {
+	e := env{}
+	e.setup(createTempDir())
+	defer e.teardown()
+
+	for _, fileName := range files {
+		t.Run(fileName, func(t *testing.T) {
+			// Get the configuration that we expect to be added.
+			expectedCfg, err := parseConfig("testdata/" + fileName)
+			if err != nil {
+				t.Fatalf("Unable to parse config: %s", fileName)
+				return
+			}
+
+			// Add the configuration.
+			if err := copyTestFile(fileName, e.fsroot); err != nil {
+				t.Fatalf("Failed adding config file: %s", fileName)
+				return
+			}
+
+			// Wait for the add to occur.
+			event := e.watch()
+
+			// Verify that the configuration was added properly.
+			if event.event != model.EventAdd {
+				t.Fatalf("Expected add event: %s", event)
+				return
```

Comment:
Here and elsewhere, no need for a return: `Fatal` exits for you (that's the primary difference between `t.Fatal` and `t.Error`: `t.Error` doesn't return).